pkg/controller/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -0,0 +1,262 @@
+	namespaceName = "toolchain-member"
+)
+
+func TestFindNamespace(t *testing.T) {
+	namespaces := []corev1.Namespace{
+		{ObjectMeta: metav1.ObjectMeta{Name: "johnsmith-dev"}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "johnsmith-code"}},
+	}
+
+	t.Run("found", func(t *testing.T) {
+		nsName := "johnsmith-dev"
+		namespace, found := findNamespace(namespaces, nsName)
+		assert.True(t, found)
+		assert.Equal(t, nsName, namespace.GetName())
+	})
+
+	t.Run("not_found", func(t *testing.T) {
+		nsName := "johnsmith-stage"
+		_, found := findNamespace(namespaces, nsName)
+		assert.False(t, found)
+	})
+}
+
+func TestNextMissingNamespace(t *testing.T) {
+	username := "johnsmith"
+	userNamespaces := []corev1.Namespace{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "johnsmith-dev", Labels: map[string]string{"revision": "rev1"},
+			},
+			Status: corev1.NamespaceStatus{Phase: corev1.NamespaceActive},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "johnsmith-code",
+			},
+			Status: corev1.NamespaceStatus{Phase: corev1.NamespaceActive},
+		},
+	}
+	tcNamespaces := []toolchainv1alpha1.Namespace{
+		{Type: "dev", Revision: "rev1"},
+		{Type: "code", Revision: "rev1"},
+		{Type: "stage", Revision: "rev1"},
+	}
+
+	t.Run("mistach_revision", func(t *testing.T) {
+		// test
+		tcNS, userNS, found := nextMissingNamespace(tcNamespaces, userNamespaces, username)
+
+		assert.True(t, found)
+		assert.Equal(t, "code", tcNS.Type)
+		assert.Equal(t, "johnsmith-code", userNS.GetName())
+	})
+
+	t.Run("found_next_namespace", func(t *testing.T) {
+		userNamespaces[1].Labels = map[string]string{"revision": "rev1"}
+
+		// test
+		tcNS, userNS, found := nextMissingNamespace(tcNamespaces, userNamespaces, username)
+
+		assert.True(t, found)
+		assert.Equal(t, "stage", tcNS.Type)
+		assert.Nil(t, userNS)
+	})
+
+	t.Run("not_found_next_namespace", func(t *testing.T) {
+		userNamespaces = append(userNamespaces, corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "johnsmith-stage", Labels: map[string]string{"revision": "rev1"},
+			},
+			Status: corev1.NamespaceStatus{Phase: corev1.NamespaceActive},
+		})
+
+		// test
+		_, _, found := nextMissingNamespace(tcNamespaces, userNamespaces, username)
+
+		assert.False(t, found)
+	})
+}
+
+func TestCreateReconcile(t *testing.T) {
+	logf.SetLogger(logf.ZapLogger(true))
+
+	username := "johnsmith"
+	nsTmplSet := newNSTmplSet(username)
+
+	reconcile := func(r *ReconcileNSTemplateSet, req reconcile.Request) {
+		res, err := r.Reconcile(req)
+		require.NoError(t, err)
+		assert.Equal(t, reconcile.Result{}, res)
+	}
+
+	t.Run("ok", func(t *testing.T) {
+		r, req, fakeClient := prepareReconcile(t, username, nsTmplSet)
+
+		// for dev
+		reconcile(r, req)
+		checkStatus(t, fakeClient, username, corev1.ConditionFalse, "Provisioning")
+
+		nsName := fmt.Sprintf("%s-dev", username)
+		activate(t, fakeClient, nsName)
+
+		reconcile(r, req)
+		checkStatus(t, fakeClient, username, corev1.ConditionFalse, "Provisioning")
```

Comment:
shouldn't we also verify that all objects were created?

Reply:
Done https://github.com/codeready-toolchain/member-operator/pull/53/commits/9e6672208631cdf61ee435dd9a7389cfa99edcbe